### ISSUE TITLE
Training time

### ISSUE THREAD

**Issue by lindemeier** [OutofAi/2D-Gaussian-Splatting#2]
Hey there,

thank you very much for this. Pretty cool.
I am thinking about using it, but the current training time is too slow. I would investigate a bit, how this can be sped up.
What are your ideas on how the training time could be reduced?

**Comment by OutofAi** [comment]
I tried to keep the whole thing simple, in terms of implementation, as much as I could and I didn't run more experiments

- You could run some ablation studies to see whether the densification makes a huge difference or whether starting with certain amount of points picked empirically for that resolution would do the job faster or the reverse starting with few points and gradually increasing whether that would make things faster
- The other thing other people did for the 3D Gaussian Splatting was to start with a lower resolution image before increasing to the a higher resolution in later epochs, reducing the initial the time needed for the first few epochs
- Currently the initialisation is relatively wrong in my code as I don't do correct inverse sigmoid for all them and I also sample randomly which is not the best solution, even in the paper for 3D Gaussian Splatting they avoided random sampling due to increase training time

**Comment by XingtongGe** [comment]
![image](https://github.com/OutofAi/2D-Gaussian-Splatting/assets/58927343/0648b420-b9cd-44e0-aee7-177f83c89b30)

I analyzed the training time and found that most of it was consumed by backpropagation. In addition, it seems that the same number of points will occupy significantly higher memory consumption compared to 3D GS. I wonder is there any way to speed up this kind of time, would it be helpful to use cuda for the rasterization and backpropagation process like in 3D GS?

**Comment by OutofAi** [comment]
I investigated a bit more, and seems like some of the issues are how pytorch handles things, mainly the decision of whether the operation is done on the cpu or gpu is not purely based on the device parameter in the code, an example of one of the bottlenecks is the einsum operation both in the forward and backward path
![image](https://github.com/OutofAi/2D-Gaussian-Splatting/assets/145302363/dfb77cd5-8fcf-4565-97dc-788d698b3296)
and if I manually do the same operation and force it on the GPU, it will speed up the process significantly for this operation
![image](https://github.com/OutofAi/2D-Gaussian-Splatting/assets/145302363/1c31857a-874d-435a-97cd-69274b079cb6)

**Comment by OutofAi** [comment]
yeah, valid points, I run more tests on it, majority of the bottleneck is either torch deciding to do an operation on cpu even though everything is on cuda or call cuda synchronisation which is causing a lock and slowdown, I think I need to do a pure cuda version as you suggested as a separate project to control the process manually rather than relying on pytorch too much. there is also a pi value that I fetch from numpy from the cpu and copy it to cuda in that code, that is not helping

**Comment by fleurs03** [comment]
> I investigated a bit more, and seems like some of the issues are how pytorch handles things, mainly the decision of whether the operation is done on the cpu or gpu is not purely based on the device parameter in the code, an example of one of the bottlenecks is the einsum operation both in the forward and backward path ![image](https://private-user-images.githubusercontent.com/145302363/297045027-dfb77cd5-8fcf-4565-97dc-788d698b3296.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgzNzY2MzksIm5iZiI6MTcxODM3NjMzOSwicGF0aCI6Ii8xNDUzMDIzNjMvMjk3MDQ1MDI3LWRmYjc3Y2Q1LThmY2YtNDU2NS05N2RjLTc4OGQ2OThiMzI5Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxNFQxNDQ1MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MDNhNDg2NDViNzI1MzdhNzcxNjkzN2RiNDZlYzZmMTE2ZDY4ZWZlYWUwZTI2MzFlZDc3OTQxYjdmNTI0YzNiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.R1OTnDQjT84_rMvPudCMMTmkfHuEGlSKd2GA2ElRLSE) and if I manually do the same operation and force it on the GPU, it will speed up the process significantly for this operation ![image](https://private-user-images.githubusercontent.com/145302363/297045539-1c31857a-874d-435a-97cd-69274b079cb6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgzNzY2MzksIm5iZiI6MTcxODM3NjMzOSwicGF0aCI6Ii8xNDUzMDIzNjMvMjk3MDQ1NTM5LTFjMzE4NTdhLTg3NGQtNDM1YS05N2NkLTY5Mjc0YjA3OWNiNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxNFQxNDQ1MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jY2FhNDNkODk4MGU1MjdlMzliN2VjZGFmYTVjNmVjODVhYmFiM2VkYTIwOWI4MmZkNDg1ODFkYmFhYTM5YzNmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.GfTCZ1I9B1THmjH5lnLWHziPelvF0lKl1FDSV8BF-pk)

Hi,

This is an interesting finding, but I'm unable to reproduce it on my end. For me, using `einsum` is more efficient in both space and time compared to breaking it down.

Here are the screenshots:

Breaking `einsum`:

<img width="1678" alt="image" src="https://github.com/OutofAi/2D-Gaussian-Splatting/assets/107610887/3f6c359b-7b5a-4b31-8e78-39c7b37fc8a1">

Using `einsum`:
<img width="1678" alt="image" src="https://github.com/OutofAi/2D-Gaussian-Splatting/assets/107610887/62014154-c89a-4214-9aec-fdd7d5fe6469">

Do you have any idea why this discrepancy might occur? It's generally expected that the official implementation of `einsum` is more efficient, which aligns with my experiments, but not with yours.

**Comment by OutofAi** [comment]
it's really difficult to pinpoint why! but if I had to guess it could either be related to pytorch version and they've recently optimised it and I was using an older version or could be the related to the amount of memory or ram that your machine utilises and its relevant speed is more efficient than mine, hence the slowdown on mine and not yours
